File: 21.合并两个有序链表.py
#
# @lc app=leetcode.cn id=21 lang=python3
#
# [21] 合并两个有序链表
#
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode):
        curr = ListNode(0)
        dummy = curr
        while not l1 and not l2:
            if l1.val < l2.val:
                curr.next = ListNode(l1.val)
                l1 = l1.next
                curr = curr.next
            else:
                curr.next = ListNode(l2.val)
                l2 = l2.next
                curr = curr.next

        while l1:
            curr.next = ListNode(l1.val)
            l1 = l1.next
            curr = curr.next
        else:
            curr.next = ListNode(l2.val)
            l2 = l2.next
            curr = curr.next

        return dummy.next


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

File: 21.合并两个有序链表.java
/*
 * @lc app=leetcode.cn id=21 lang=java
 *
 * [21] 合并两个有序链表
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode res = new ListNode(0);
        ListNode dummy = res;
        // 拿小值
        // 当前存起来next赋值
        // 存起来
        while (l1 != null && l2 != null) {
            if (l1.val < l2.val) {
                res.next = new ListNode(l1.val);
                l1 = l1.next;
            } else {
                res.next = new ListNode(l2.val);
                l2 = l2.next;
            }
            res = res.next;
        }
        while (l1 != null) {
            res.next = l1;
            l1 = l1.next;
            res = res.next;
        }
        while (l2 != null) {
            res.next = l2;
            l2 = l2.next;
            res = res.next;
        }
        return dummy.next;
    }
}

